cmd/libs/go2idl/make-client/generators/client-generator.go

```diff
@@ -0,0 +1,283 @@
+		// single file.
+		GeneratorFunc: func(c *generator.Context) (generators []generator.Generator) {
+			generators = []generator.Generator{
+				// Always generate a "doc.go" file.
+				generator.DefaultGen{OptionalName: "doc"},
+			}
+			// Since we want a file per type that we generate a set for, we
+			// have to provide a function for this.
+			for _, t := range c.Order {
+				generators = append(generators, &genClient{
+					DefaultGen: generator.DefaultGen{
+						// Use the privatized version of the
+						// type name as the file name.
+						//
+						// TODO: make a namer that converts
+						// camelCase to '-' separation for file
+						// names?
+						OptionalName: c.Namers["private"].Name(t),
+					},
+					outputPackage: arguments.OutputPackagePath,
+					typeToMatch:   t,
+					imports:       generator.NewImportTracker(),
+				})
+			}
+			return generators
+		},
+		FilterFunc: func(c *generator.Context, t *types.Type) bool {
+			_, found := topLevelTypes[t]
+			return found
+		},
+	}}
+}
+
+// genClient produces a file with a set for a single type.
+type genClient struct {
+	generator.DefaultGen
+	outputPackage string
+	typeToMatch   *types.Type
+	imports       *generator.ImportTracker
+}
+
+// Filter ignores all but one type because we're making a single file per type.
+func (g *genClient) Filter(c *generator.Context, t *types.Type) bool { return t == g.typeToMatch }
+
+func (g *genClient) Namers(c *generator.Context) namer.NameSystems {
+	return namer.NameSystems{
+		"raw": namer.NewRawNamer(g.outputPackage, g.imports),
+	}
+}
+
+func (g *genClient) Imports(c *generator.Context) (imports []string) {
+	return append(g.imports.ImportLines())
+}
+
+// args constructs arguments for templates. Usage:
+// g.args(t, "key1", value1, "key2", value2, ...)
+//
+// 't' is loaded with the key 'type'.
+//
+// We could use t directly as the argument, but doing it this way makes it easy
+// to mix in additional parameters. This feature is not used in this set
+// generator, but is present as an example.
+func (g *genClient) args(t *types.Type, kv ...interface{}) interface{} {
+	m := map[interface{}]interface{}{"type": t}
+	// @lavalamp, shall I create a namer?
+	pkg := filepath.Base(t.Name.Package)
+	m["package"] = pkg
+	m["Package"] = namer.IC(pkg)
+	// @lavalamp: how about making a plural function in pkg/util, RESTMapper also needs such a function.
+	// m["plural"] = util.Plural(type.Name.Name)
+	m["fieldSelector"] = fieldSelector
+	m["labelSelector"] = labelSelector
+	m["watchInterface"] = watchInterface
+	m["apiDeleteOptions"] = apiDeleteOptions
+	m["apiListOptions"] = apiListOptions
+	for i := 0; i < len(kv)/2; i++ {
+		m[kv[i*2]] = kv[i*2+1]
+	}
+	return m
+}
+
+// GenerateType makes the body of a file implementing a set for type t.
+func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	sw.Do(nameSpacerTemplate, g.args(t))
```

Comment:
Functions are injected in the .Do method. There's no benefit to parsing
early since they won't break once you get them working, and they're not
reused, and go2idl is not slow anyway.

On Fri, Nov 20, 2015 at 2:23 PM, krousey <notifications@github.com> wrote:

> In cmd/libs/go2idl/make-client/generators/client-generator.go
> <https://github.com/kubernetes/kubernetes/pull/17534#discussion_r45526693>
> :
>
> > +	// m["plural"] = util.Plural(type.Name.Name)
> > +	m["fieldSelector"] = fieldSelector
> > +	m["labelSelector"] = labelSelector
> > +	m["watchInterface"] = watchInterface
> > +	m["apiDeleteOptions"] = apiDeleteOptions
> > +	m["apiListOptions"] = apiListOptions
> > +	for i := 0; i < len(kv)/2; i++ {
> > +		m[kv[i*2]] = kv[i*2+1]
> > +	}
> > +	return m
> > +}
> > +
> > +// GenerateType makes the body of a file implementing a set for type t.
> > +func (g *genClient) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
> > +	sw := generator.NewSnippetWriter(w, c, "$", "$")
> > +	sw.Do(nameSpacerTemplate, g.args(t))
>
> Why are we waiting until this point to try to parse the templates? I would
> think it'd be better to use template.MustParse at the package level
> definition so you wouldn't have to have a unit test hit these lines to know
> you have template errors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17534/files#r45526693>.
>